barretenberg/cpp/src/barretenberg/commitment_schemes/ipa/ipa.fuzzer.cpp

```diff
@@ -0,0 +1,158 @@
+#include "barretenberg/srs/factories/file_crs_factory.hpp"
+
+namespace bb {
+
+// We actually only use 4, because fuzzing is very slow
+constexpr size_t COMMITMENT_TEST_NUM_POINTS = 32;
+using Curve = curve::Grumpkin;
+std::shared_ptr<CommitmentKey<Curve>> ck;
+std::shared_ptr<VerifierCommitmentKey<Curve>> vk;
+/**
+ * @brief Class that allows us to call internal IPA methods, because it's friendly
+ *
+ */
+class ProxyCaller {
+  public:
+    template <typename Transcript>
+    static void compute_opening_proof_internal(const std::shared_ptr<CommitmentKey<Curve>>& ck,
+                                               const OpeningPair<Curve>& opening_pair,
+                                               const Polynomial<Curve::ScalarField>& polynomial,
+                                               const std::shared_ptr<Transcript>& transcript)
+    {
+        IPA<Curve>::compute_opening_proof_internal(ck, opening_pair, polynomial, transcript);
+    }
+    template <typename Transcript>
+    static bool verify_internal(const std::shared_ptr<VerifierCommitmentKey<Curve>>& vk,
+                                const OpeningClaim<Curve>& opening_claim,
+                                const std::shared_ptr<Transcript>& transcript)
+    {
+        return IPA<Curve>::verify_internal(vk, opening_claim, transcript);
+    }
+};
+} // namespace bb
+
+/**
+ * @brief Initialize SRS, commitment key, verification key
+ *
+ */
+extern "C" void LLVMFuzzerInitialize(int*, char***)
+{
+    srs::init_grumpkin_crs_factory("../srs_db/ignition");
+    ck = std::make_shared<CommitmentKey<Curve>>(COMMITMENT_TEST_NUM_POINTS);
+    auto crs_factory = std::make_shared<srs::factories::FileCrsFactory<curve::Grumpkin>>("../srs_db/grumpkin",
+                                                                                         COMMITMENT_TEST_NUM_POINTS);
+    vk = std::make_shared<VerifierCommitmentKey<curve::Grumpkin>>(COMMITMENT_TEST_NUM_POINTS, crs_factory);
+}
+
+// This define is needed to make ProxyClass a friend of IPA
+#define IPA_FUZZ_TEST
+#include "ipa.hpp"
+
+/**
+ * @brief A fuzzer for the IPA primitive
+ *
+ * @details Parses the given data as a polynomial, a sequence of challenges for the transcript and the evaluation point,
+ * then opens the polynomial with IPA and verifies that the opening was correct
+ */
+extern "C" int LLVMFuzzerTestOneInput(const unsigned char* data, size_t size)
+{
+    using Fr = grumpkin::fr;
+    using Polynomial = Polynomial<Fr>;
+    // We need data
+    if (size == 0) {
+        return 0;
+    }
+    // Get the logarighmic size of polynomial
+    const auto log_size = static_cast<size_t>(data[0]);
+    // More than 4 is so bad
+    if (log_size == 0 || log_size > 2) {
+        return 0;
+    }
+    const auto* offset = data + 1;
+    const auto num_challenges = log_size + 1;
+    // How much data do we need?
+    // Challenges: sizeof(uint256_t) * num_challenges + 1 for montgomery switch
```

Comment:
wdym by montgomery switch?

Reply:
When we do field(uint256_t(buffer)), underneath the value converts to montgomery form, after it is reduced several times. This means that the internal data of the field element  is a map of the buffer, and so a single bit change of the buffer will lead to huge changes in internal representation. To be able to switch it off, I use flags from input to convert the field element back from montgomery form, so the internal form directly corresponds with what's in the buffer

Reply:
thannks for explaining!